### ISSUE TITLE
Duplicate records created when using batch_size in exportBulkRecords

### ISSUE THREAD

**Issue by harrelfe** [vubiostat/redcapAPI#262]
I'm running on the ACTIV6 EDC.
```
d <- exportBulkRecords(list(D=edc_conn),
                       forms=list(D='demographics'),
                       batch_size=4735,
                       post = function(Records, rcon) {
                         Records |> mChoiceCast(rcon) |>
                           guessDate(rcon, threshold=0.6)})
dem1 <- d$D_demographics
r <- dem1$record_id
sort(r[duplicated(r)])
```

When `batch_size` is removed I get fewer records, i.e., no duplicates.

**Comment by spgarbet** [comment]
We need to determine the source of duplicates. I.e., is is a bug in this library or is it a problem with the data coming from the server when using this parameter.

**Comment by nutterb** [comment]
@harrelfe , by any chance, are there multiple records per subject in the demographics form?

@spgarbet , the reason I ask this is related to the intended behavior of batching the records. The process is supposed to be

1. Export just the record IDs 
2. Group the record IDs into batches of batch_size
3. For each batch, export the records associated with the batch of record IDs.

the implementation at https://github.com/vubiostat/redcapAPI/blob/main/R/exportRecordsTyped.R#L677-L681 doesn't look like it's accounting for the possibility of multiple records. So if we have a batch size of 5 and the export in step one produces

| record_id | 
|-----------|
| 1 | 
| 2 | 
| 3 | 
| 4 | 
| 5 | 
| 5 | 
| 6 | 
| 7 | 
| 8 |
| 9 |

We will get two batches of {1, 2, 3, 4, 5} and {5, 6, 7, 8 9}.  The records for ID 5 will get exported twice.

If this is the case, then the solution might be as simple as putting `records <- unique(records[[target_field]])` on line 681.

I hope that's an intelligible description.

**Comment by nutterb** [comment]
Ah, just figured it out.  The conditions for this are created when a record ID has data recorded over multiple events. When the record IDs are pulled from the API, it returns something like 

| record_id | redcap_event_name | 
|-----------|------------------------|
| 1             | event_1_arm_1          | 
| 2             | event_1_arm_1          | 
| 2             | event_2_arm_1          | 

In this case, if the `batch_size = 2`, the two batches will be {1, 2} and {2}.  The records for id 2 will be exported twice, resulting in the duplicates.

**Comment by harrelfe** [comment]
The only part I don't get is that I'm dealing with baseline-only datasets where there are no record_id duplicates when batch_size is not specified.

**Comment by nutterb** [comment]
I was able to engineer a scenario that reproduces the record duplication. 

Would you be willing to run `devtools::install_github("redcapAPI", ref = "issue-262-duplicate-bulk-records")` and then run your code from your initial posting? I believe that patch will avoid the duplication.

**Comment by spgarbet** [comment]
I've merged into the main branch of 2.8.1

**Comment by harrelfe** [comment]
I re-ran the API with the new version and no duplicates.  Thanks!
